From 34b3596521115bf8ec1655c461c5a3190887f6e8 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 25 Apr 2020 15:21:24 +0200
Subject: [PATCH 0612/1013] evl/work: add safe async call form

We need a safe form of async workers to address the following issue:

CPU0: evl_call_inband*(arg=some_element)
CPU0:     irq_work_queue()
CPU0:          queue_work()
CPU1: release(element)
CPU1:     dispose(element)
CPU0:             work->handler(arg=some_element)
CPU0:                 invalid access to stale 'some_element'

evl_init_work_safe() guarantees integrity by holding a reference on
the given element from when evl_call_inband*() is issued until the
handler has completed.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/evl/work.h |  8 ++++++++
 kernel/evl/work.c  | 34 ++++++++++++++++++++++++++++++++--
 2 files changed, 40 insertions(+), 2 deletions(-)

diff --git a/include/evl/work.h b/include/evl/work.h
index 32f2e4a66290..e3cc4a2a9409 100644
--- a/include/evl/work.h
+++ b/include/evl/work.h
@@ -11,11 +11,14 @@
 #include <linux/workqueue.h>
 #include <evl/flag.h>
 
+struct evl_element;
+
 struct evl_work {
 	struct irq_work irq_work;
 	struct work_struct wq_work;
 	struct workqueue_struct *wq;
 	int (*handler)(void *arg);
+	struct evl_element *element;
 };
 
 struct evl_sync_work {
@@ -27,6 +30,10 @@ struct evl_sync_work {
 void evl_init_work(struct evl_work *work,
 		   void (*handler)(struct evl_work *work));
 
+void evl_init_work_safe(struct evl_work *work,
+			void (*handler)(struct evl_work *work),
+			struct evl_element *element);
+
 void evl_init_sync_work(struct evl_sync_work *sync_work,
 			int (*handler)(struct evl_sync_work *sync_work));
 
@@ -41,6 +48,7 @@ static inline void evl_call_inband(struct evl_work *work)
 static inline
 void evl_flush_work(struct evl_work *work)
 {
+	irq_work_sync(&work->irq_work);
 	flush_work(&work->wq_work);
 }
 
diff --git a/kernel/evl/work.c b/kernel/evl/work.c
index cb14c6840316..eb7eef177ada 100644
--- a/kernel/evl/work.c
+++ b/kernel/evl/work.c
@@ -4,6 +4,7 @@
  * Copyright (C) 2020 Philippe Gerum  <rpm@xenomai.org>
  */
 
+#include <evl/factory.h>
 #include <evl/work.h>
 
 static void do_wq_work(struct work_struct *wq_work)
@@ -12,6 +13,8 @@ static void do_wq_work(struct work_struct *wq_work)
 
 	work = container_of(wq_work, struct evl_work, wq_work);
 	work->handler(work);
+	if (work->element)
+		evl_put_element(work->element);
 }
 
 static void do_wq_work_sync(struct work_struct *wq_work)
@@ -28,7 +31,9 @@ static void do_irq_work(struct irq_work *irq_work)
 	struct evl_work *work;
 
 	work = container_of(irq_work, struct evl_work, irq_work);
-	queue_work(work->wq, &work->wq_work);
+
+	if (!queue_work(work->wq, &work->wq_work) && work->element)
+		evl_put_element(work->element);
 }
 
 void evl_init_work(struct evl_work *work,
@@ -37,9 +42,19 @@ void evl_init_work(struct evl_work *work,
 	init_irq_work(&work->irq_work, do_irq_work);
 	INIT_WORK(&work->wq_work, do_wq_work);
 	work->handler = (typeof(work->handler))handler;
+	work->element = NULL;
 }
 EXPORT_SYMBOL_GPL(evl_init_work);
 
+void evl_init_work_safe(struct evl_work *work,
+			void (*handler)(struct evl_work *work),
+			struct evl_element *element)
+{
+	evl_init_work(work, handler);
+	work->element = element;
+}
+EXPORT_SYMBOL_GPL(evl_init_work_safe);
+
 void evl_init_sync_work(struct evl_sync_work *sync_work,
 			int (*handler)(struct evl_sync_work *sync_work))
 {
@@ -48,6 +63,11 @@ void evl_init_sync_work(struct evl_sync_work *sync_work,
 	init_irq_work(&work->irq_work, do_irq_work);
 	INIT_WORK(&work->wq_work, do_wq_work_sync);
 	work->handler = (typeof(work->handler))handler;
+	/*
+	* No point in holding a safety reference since the caller
+	* waits for the handler to have returned.
+	*/
+	work->element = NULL;
 	evl_init_flag(&sync_work->done);
 }
 EXPORT_SYMBOL_GPL(evl_init_sync_work);
@@ -56,7 +76,17 @@ void evl_call_inband_from(struct evl_work *work,
 			struct workqueue_struct *wq)
 {
 	work->wq = wq;
-	irq_work_queue(&work->irq_work);
+
+	/*
+	 * Async call: get a reference on the protected element if
+	 * specified so that the handler may access such element
+	 * safely after our caller has unwound.
+	 */
+	if (work->element)
+		evl_get_element(work->element);
+
+	if (!irq_work_queue(&work->irq_work) && work->element)
+		evl_put_element(work->element);
 }
 EXPORT_SYMBOL_GPL(evl_call_inband_from);
 
-- 
2.30.2

